File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.py
from itertools import combinations, permutations

def isPrime(number):
    if number < 2: return False
    for i in range(2, int(number ** 0.5) + 1):
        if number % i == 0: return False
    return True

def solution(numbers):
    answer = set()
    for i in range(1, len(numbers) + 1):
        for c in list(combinations(numbers, i)):
            for p in list(permutations(c, i)):
                number = int(''.join(p))
                if number >= 2 and isPrime(number): answer.add(number)
                
    return len(answer)

File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.java
// (number, numbers) -> number로 시작해서 numbers로 만들 수 있는 숫자 중 소수의 집합
import java.util.*;
import java.util.stream.*;
class Solution {
    static int N;
    public int solution(String nums) {
        Set<Integer> primes = new HashSet<>();
        int[] numbers = nums.chars().map(n -> n - '0').toArray();
        N = numbers.length;
        subset(0, numbers, new boolean[N], primes);
        return primes.size();
    }
    
    public boolean isPrime(int number){
        if(number <= 1) return false;
        if(number == 2 || number == 3) return true;
        for(int i = 2; i < Math.sqrt(number) + 1; i++){
            if(number % i == 0) return false;
        }
        return true;
    }
    
    public void subset(int number, int[] numbers, boolean[] isSelected, Set<Integer> primes){
        if(isPrime(number)) primes.add(number);
        
        for(int i = 0; i < numbers.length; i++){
            if(isSelected[i]) continue;
            int next = number * 10 + numbers[i];
            
            isSelected[i] = true;
            subset(next, numbers, isSelected, primes);
            isSelected[i] = false;
        }
    }
}